### PR TITLE
feat: support more elementwise and unary dynamo converters

### LINKED ISSUE #2208: Expose IElementWiseLayer in dynamo.conversion.impl

### PR DESCRIPTION
**Pull request by zewenli98**
# Description

Support more elementwise and unary dynamo converters.

For `bitwise_op` converters, TensorRT doesn't support bitwise ops. Thus, if inputs are `int` values, they will fall back to pytorch. If inputs are `bool`, they will be converted with the help of `logical_op`.

Fixes #2208 #2199 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/ops.py [hunk 1]
```diff
@@ -422,13 +423,46 @@ def logical_xor(
...
+
 def eq(
     ctx: ConversionContext,
     target: Target,
     source_ir: Optional[SourceIR],
     name: str,
-    lhs_val: Union[TRTTensor, int, float],
-    rhs_val: Union[TRTTensor, int, float],
+    lhs_val: TRTTensor,
+    rhs_val: Union[TRTTensor, int, float, bool, Sequence[Union[int, float, bool]]],
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/elementwise/ops.py:465]: In what cases could this be a sequence of integers, floats, or bools? Could the `convert_binary_elementwise` function handle such situations?

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/ops.py [hunk 2]
```diff
@@ -441,13 +475,30 @@ def eq(
...
+
 def gt(
     ctx: ConversionContext,
     target: Target,
     source_ir: Optional[SourceIR],
     name: str,
-    lhs_val: Union[TRTTensor, int, float],
-    rhs_val: Union[TRTTensor, int, float],
+    lhs_val: TRTTensor,
+    rhs_val: Union[TRTTensor, int, float, Sequence[Union[int, float]]],
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/elementwise/ops.py:501]: As mentioned above, is it necessary to have this sequence case? It does not appear in the schemas, [as here](https://github.com/pytorch/pytorch/blob/d977f118ad5422bec0b57f5b98d7124a6e2366a3/aten/src/ATen/native/native_functions.yaml#L8817).

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 3]
```diff
@@ -1647,9 +1647,128 @@ def aten_ops_logical_xor(
...
+
+    if lhs_meta is None or rhs_meta is None:
+        return False
+
+    supported_type = [torch.bool, bool]
+    return lhs_meta.dtype in supported_type and rhs_meta.dtype in supported_type
+
+
+@dynamo_tensorrt_converter(torch.ops.aten.bitwise_and.Tensor, capability_validator=bitwise_type_validator)  # type: ignore[misc]
+@dynamo_tensorrt_converter(torch.ops.aten.bitwise_and.Scalar)  # type: ignore[misc]
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:1672]: The `# type: ignore[misc]` can be removed as discussed

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 4]
```diff
@@ -1647,9 +1647,128 @@ def aten_ops_logical_xor(
...
+        torch.ops.aten.bitwise_or.Tensor,
+        torch.ops.aten.bitwise_xor.Tensor,
+    ]
+    if node.target not in targets:
+        return False
+
+    lhs_val = node.args[0]
+    rhs_val = node.args[1]
+    lhs_meta = lhs_val.meta.get("tensor_meta")
+    rhs_meta = rhs_val.meta.get("tensor_meta")
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:1662]: If `rhs` is a Scalar, does this error out?

**Review by gs-olive — COMMENTED** [review]
Overall looks good - had a few questions and suggestions as mentioned in the comments

File: py/torch_tensorrt/dynamo/conversion/impl/elementwise/ops.py (hunk 1 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/elementwise/ops.py]: Thanks for the comments. At first I thought it could be a sequence of integers, floats, or bools, but I checked the doc it should not be.

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 4 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: I added support for Scalar with bool type, and related tests.

File: tests/py/dynamo/conversion/test_eq_aten.py [hunk 5]
```diff
@@ -2,64 +2,65 @@
...
     @parameterized.expand(
         [
-            ("2d", (2, 1), 1),
-            ("3d", (2, 1, 2), 2.0),
+            ("2d", (5, 3), 1),
+            ("3d", (5, 3, 2), 2.0),
         ]
     )
-    def test_equal_tensor_scalar(self, _, shape, scalar):
-        class equal(nn.Module):
```
Comment (apbose) [inline tests/py/dynamo/conversion/test_eq_aten.py:36]: Is this case different from the previous `torch.ops.aten.eq.Tensor` case? As in in terms of the trace? Same application for the other `ge`, `gt`, `le`, `lt`, `ne`?

File: tests/py/dynamo/conversion/test_bitwise_and_aten.py
Comment (apbose) [inline tests/py/dynamo/conversion/test_bitwise_and_aten.py:1]: Overall everything and the test cases look good to me. Just one suggestion, can all the similar test cases files be consolidated into one, using lambda functions like `test_binary_ops_aten.py`? Since the structure of the test classes are similar for `and`, `or` and `xor`. Similarly for` eq`, `ge`, `gt`,` le`,` lt`,` ne`

File: tests/py/dynamo/conversion/test_eq_aten.py (hunk 5 above)
Comment (zewenli98) [reply tests/py/dynamo/conversion/test_eq_aten.py:36]: I think `torch.ops.aten.eq.Tensor` and `torch.ops.aten.eq.Scalar` are different. The main difference is the second argument. In `torch.ops.aten.eq.Tensor`, it is `Tensor` while in `torch.ops.aten.eq.Scalar`, it's `Scalar`.

File: tests/py/dynamo/conversion/test_bitwise_and_aten.py
Comment (zewenli98) [reply tests/py/dynamo/conversion/test_bitwise_and_aten.py:1]: Well, that's a good point. Since the previous tests are separated into many files even though they share the similarity (like `unary` or `elementwise`), as you said, I just follow the convention. But I do agree with you that the current tests are kind of redundant, so maybe we can refactor them later?

**Review by gs-olive — COMMENTED** [review]
Looks good to me overall! One question - should `convert_binary_elementwise` also be allowed to handle `bool` types?

**Comment by zewenli98** [comment]
> Looks good to me overall! One question - should `convert_binary_elementwise` also be allowed to handle `bool` types?

Yes, I agree with you. Fixed! Thanks for the comments.

**Review by apbose — APPROVED** [review]

**Review by gs-olive — COMMENTED** [review]
Just a minor comment, `bool` should likely be added here, to lines 123 + 127 since this PR adds support for `bool` in `convert_binary_elementwise`:
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py#L123-L130
Otherwise, seems good to go!
